### PR TITLE
fix(governance-store): emit op-events after the op-log persists (#2770)

### LINKED ISSUE #2770: governance-store: OpEvents emitted before op-log entry is persisted (subscriber read-back race)
## Problem

In the group-op apply path, `notify_op_event(...)` fires **inside the per-op handlers** (e.g. `OpEvent::TeeMemberAdmitted` at `crates/governance-store/src/ops/group/member_joined_via_tee_attestation.rs:43`) **before** the caller persists the op-log entry (`crates/governance-store/src/lib.rs:~1340` / `crates/governance-store/src/namespace/governance.rs:~1365`). There is no transaction wrapping the apply — writes are write-through (`crates/store/src/layer.rs:93-95`), and `notify` dispatches to subscribers on other tasks.

Result: a subscriber that reacts to an event and immediately reads the **op-log** (e.g. `tee_admission_record`, `is_tee_admitted_identity`, `read_tee_admission_policy`, `read_op_log_after`) can observe stale state (entry not yet written). Membership rows are written before the event, so `has_direct_member`/`list` are unaffected — only op-log-scanning reads race.

This currently affects `calimero-context::tee_subgroup_admit` (`handle_new_tee_member` reads back the verdict via `tee_admission_record` on `TeeMemberAdmitted`). It is mitigated there by a bounded wake-then-reread (commit e2b12885), but the underlying emit-before-persist ordering is latent for any future op-log-reading subscriber, and is already noted as a deferred "single atomic batch" item at `crates/governance-store/src/local_state.rs:223-226`.

## Proposed durable fix (Option A)

Collect events during apply and flush them **after** the op-log entry persists (ideally once the apply uses a single atomic `StoreBatch`/`Store::apply`). Add a `pending_events: Vec<OpEvent>` to the apply ctx (mirroring the existing `divergence` field), have per-op handlers queue instead of `notify`, and drain after persist. Gate the flush on the op-log-append path so idempotent replays/dedup don't re-emit. All current subscribers (`self_purge`, `auto_follow`, `tee_subgroup_admit`) are idempotent, so emit-on-append-only is the safer, more correct behavior — but verify against the e2e suite.

Blast radius: ~8 files in `calimero-governance-store` (per-op handlers + two flush sites). `GroupKeyDelivered` (join_group wake-then-reread) must be left as-is. No public-API change.

## Context
Found during Phase 1 of transparent per-subgroup TEE admission. The wake-then-reread in `tee_subgroup_admit` makes that feature correct meanwhile; this issue tracks the root-cause fix.

### PR DESCRIPTION
**Pull request by xilosada**
## Summary

Fixes #2770: governance `OpEvent`s were emitted **before** the op-log entry persisted, so a subscriber reacting to an event and reading the op-log back could race the write. This is what forced the bounded wake-then-reread workaround in the per-subgroup TEE admission subscriber and opened a "self_purge deletes the op-log while apply is still writing it" window.

**Fix — collect-then-flush.** Each apply context (`GroupApplyCtx`, `NamespaceApplyCtx`) gains a `pending_events` sink + `queue_event`, mirroring the existing `divergence` out-parameter. Per-op handlers **queue** events instead of calling `op_events::notify`; the apply functions **drain + notify only after the op-log append succeeds**, on both apply paths (group-op and namespace RootOp).

- **Emit-on-append-only:** drains live only on the append branch, never the content-hash dedup early-return — so a replayed/duplicate op no longer re-emits (all subscribers are idempotent + lossy-tolerant, so this is strictly more correct). Documented behavior change.
- **`GroupKeyDelivered` is intentionally left a direct `notify`** — it's off the op-append path and its `join_group` wake is latency-sensitive.
- Event ordering preserved (`MemberRemoved` before `TeeMemberRemoved`) via FIFO queue→drain.

This is **PR-1 of a 3-PR set** for TEE-lifecycle correctness (shared design: `docs/superpowers/specs/2026-06-17-tee-lifecycle-correctness-design.md`). It is independent and mergeable on its own. PR-2 (#2726, subgroup purge reconcile) depends on #2776; PR-3 (#2771, born-Open create) depends on #2772.

## Test Plan
- [x] Group race-gone: a concurrent observer confirms the op-log entry is persisted at the instant `MemberAdded` fires (verified RED before the fix, GREEN after).
- [x] Namespace race-gone: same concurrent-observer proof for `SubgroupCreated` vs the namespace op persist.
- [x] Replay-no-re-emit: a re-applied (already-logged) op fires no event.
- [x] `cargo test -p calimero-governance-store -p calimero-context -p calimero-node` green; `cargo fmt --all --check` clean (1.88.0). Existing subscribers (`auto_follow`, `self_purge`) unaffected.

## Scope
Internal to `calimero-governance-store` only — no `calimero-server-primitives`/`calimero-tee-attestation` public-API change, no mero-tee rev bump. The `tee_subgroup_admit` wake-then-reread mitigation is intentionally **not** removed here (it lives in #2772; its removal is a follow-up once both land).

## Follow-up note
Replays no longer re-emit op-events (idempotent subscribers unaffected).

---

> [!NOTE]
> **Medium Risk**
> Changes ordering and replay semantics on the governance apply hot path (group + namespace), which `auto_follow` and `self_purge` depend on; behavior is intentional, documented, and covered by new tests, with a bounded crash-window edge case accepted.
> 
> **Overview**
> Fixes **#2770** by changing when governance **`OpEvent`s** are delivered: handlers no longer call `op_events::notify` during apply.
> 
> **Collect-then-flush.** `GroupApplyCtx` and `NamespaceApplyCtx` gain a `pending_events` queue (same pattern as `divergence`). Group and namespace op handlers **`queue_event`** instead of notifying. **`apply_local_signed_group_op`**, namespace **`apply_group_op_inner`**, and **`apply_signed_op`** (RootOp path) **drain and notify only after** the op-log / namespace op is appended.
> 
> **Emit-on-append-only:** On content-hash or namespace dedup early-return, queued events are **dropped** so replays and duplicates do **not** re-emit (documented tradeoff vs a rare crash-between-append-and-flush gap). **`build_auto_follow_set_if_enabled`** (renamed from `emit_auto_follow_set_if_enabled`) returns `Option<OpEvent>` for the queue. **`GroupKeyDelivered`** stays a direct notify (unchanged).
> 
> Tests add concurrent-observer checks that the log contains the op when events fire, plus replay no re-emit; design/plan docs are added under `docs/superpowers/`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 138470b71a4dc511229b9185b282055a9b16b27c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: crates/governance-store/src/lib.rs [hunk 1]
```diff
@@ -1347,6 +1353,10 @@ pub fn apply_local_signed_group_op(store: &Store, op: &SignedGroupOp) -> EyreRes
         &op_bytes,
     )?;
 
+    // #2770: flush events only after the op-log entry is durably appended.
+    for event in pending_events {
+        crate::op_events::notify(event);
+    }
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/lib.rs:1397]: Rejecting the autofix (re-emit on the dedup branch) — it conflicts with this PR's core intent.

The dedup branch cannot distinguish a crash-recovery replay from an ordinary duplicate delivery (network re-gossip, DAG replay). Re-emitting there fires the event again on **every** normal replay, reintroducing exactly the duplicate-notification behaviour this PR removes, and inverting the `replayed_group_op_does_not_re_emit` test.

The crash window you describe (crash between the op-log append and the post-append flush → one-time signal like `TeeMemberRemoved` not redelivered) is real but is the **same bounded, already-accepted gap** documented for the broadcast lagged-drop case in `calimero-context::self_purge` (run-loop `Lagged` arm, self_purge.rs:147-188): it is **not** a forward-secrecy hole — FS on future writes is held by the key-rotation pipeline, not this purge — and the residue is only stale, already-orphaned local key material.

A genuinely crash-safe fix is to write the self-purge marker **deterministically on the apply path** (every node, independent of event delivery) rather than in the event handler — that is TEE-lifecycle follow-up territory (#2772/#2771), out of scope here.

Documented the tradeoff in code: a canonical note on the dedup branch in `apply_local_signed_group_op` (lib.rs), referenced from the namespace group-op and RootOp dedup branches. Pushed in ff54cb0d.

File: crates/governance-store/src/tests.rs [hunk 2]
```diff
@@ -5749,6 +5751,271 @@ mod auto_follow_tests {
...
+        );
+    }
+
+    /// #2770: re-applying an already-logged op must NOT re-fire its event.
+    /// The queued events are dropped on the content-hash dedup early-return.
+    #[test]
+    fn replayed_group_op_does_not_re_emit() {
+        use crate::op_events::{self, OpEvent};
+
+        let mut rng = OsRng;
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5894]: Replaced the 300ms wall-clock window with a synchronous single-pass `try_recv` drain (no sleep). `apply_local_signed_group_op` and `op_events::notify` are both synchronous on the test thread, so any re-emitted event is already buffered when the replay apply returns — a buggy re-emit now fails deterministically instead of racing a deadline. Pushed in ff54cb0d.

File: crates/governance-store/src/tests.rs [hunk 3]
```diff
@@ -5749,6 +5751,271 @@ mod auto_follow_tests {
...
+        let (store, _gid, gid_bytes, admin_sk, _existing_member_sk) = seed(&mut rng);
+
+        let new_member_sk = PrivateKey::random(&mut rng);
+        let new_member_pk = new_member_sk.public_key();
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5898]: Replaced the 300ms wall-clock window with a synchronous single-pass `try_recv` drain (no sleep). `apply_local_signed_group_op` and `op_events::notify` are both synchronous on the test thread, so any re-emitted event is already buffered when the replay apply returns — a buggy re-emit now fails deterministically instead of racing a deadline. Pushed in ff54cb0d.

File: crates/governance-store/src/tests.rs [hunk 4]
```diff
@@ -5749,6 +5751,271 @@ mod auto_follow_tests {
...
+            log_contained_at_emit,
+            "op-log entry must be persisted before MemberAdded fires (#2770)"
+        );
+    }
+
+    /// #2770: re-applying an already-logged op must NOT re-fire its event.
+    /// The queued events are dropped on the content-hash dedup early-return.
+    #[test]
+    fn replayed_group_op_does_not_re_emit() {
+        use crate::op_events::{self, OpEvent};
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5892]: Replaced the 300ms wall-clock window with a synchronous single-pass `try_recv` drain (no sleep). `apply_local_signed_group_op` and `op_events::notify` are both synchronous on the test thread, so any re-emitted event is already buffered when the replay apply returns — a buggy re-emit now fails deterministically instead of racing a deadline. Pushed in ff54cb0d.

File: crates/governance-store/src/tests.rs [hunk 5]
```diff
@@ -5749,6 +5751,271 @@ mod auto_follow_tests {
...
+                member: new_member_pk,
+                role: GroupMemberRole::Member,
+            },
+        )
+        .unwrap();
+        let content_hash = op.content_hash().unwrap();
+
+        // Observer thread: blocks on `recv()`, then snapshots whether the
+        // op-log already contains our op at the instant the event fires.
+        // `Store` is `Clone` (shared handle to the same backing DB), so the
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5805]: Both op-ordering observers now rendezvous via `std::sync::Barrier::new(2)` instead of the pre-loop ready-signal, so the observer is provably in its `try_recv` spin loop before the applier proceeds (closes the vacuous-pass window). Also handled `TryRecvError::Lagged(n)` explicitly: it now `panic!`s with a clear message rather than falling into `Err(_) => return None`, which the caller would have misreported as "event never fired". Pushed in ff54cb0d.

[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5805]: Both op-ordering observers now rendezvous via `std::sync::Barrier::new(2)` instead of the pre-loop ready-signal, so the observer is provably in its `try_recv` spin loop before the applier proceeds (closes the vacuous-pass window). Also handled `TryRecvError::Lagged(n)` explicitly: it now `panic!`s with a clear message rather than falling into `Err(_) => return None`, which the caller would have misreported as "event never fired". Pushed in ff54cb0d.

[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5805]: Both op-ordering observers now rendezvous via `std::sync::Barrier::new(2)` instead of the pre-loop ready-signal, so the observer is provably in its `try_recv` spin loop before the applier proceeds (closes the vacuous-pass window). Also handled `TryRecvError::Lagged(n)` explicitly: it now `panic!`s with a clear message rather than falling into `Err(_) => return None`, which the caller would have misreported as "event never fired". Pushed in ff54cb0d.

File: crates/governance-store/src/ops/namespace/context.rs [hunk 6]
```diff
@@ -34,6 +40,14 @@ impl<'a> NamespaceApplyCtx<'a> {
         self.namespace_id
     }
 
+    pub(crate) fn queue_event(&mut self, event: crate::op_events::OpEvent) {
+        self.pending_events.push(event);
+    }
+
+    pub(crate) fn take_events(&mut self) -> Vec<crate::op_events::OpEvent> {
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/ops/namespace/context.rs:50]: Added a doc comment marking `take_events` as a destructive drain (leaves the buffer empty; a second call returns an empty Vec; the post-persist flush must take exactly once). Pushed in ff54cb0d.

File: crates/governance-store/src/tests.rs [hunk 7]
```diff
@@ -5749,6 +5751,301 @@ mod auto_follow_tests {
...
+        assert!(
+            log_contained_at_emit,
+            "op-log entry must be persisted before MemberAdded fires (#2770)"
+        );
+    }
+
+    /// #2770: re-applying an already-logged op must NOT re-fire its event.
+    /// The queued events are dropped on the content-hash dedup early-return.
+    #[test]
+    fn replayed_group_op_does_not_re_emit() {
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5891]: Good catch — applied. The drain `Err(_) => break` swallowed `Lagged` the same as `Empty`, so a capacity flood between subscribe and drain could have dropped a buggy re-emit and passed vacuously. It now `panic!`s on `Lagged(n)` (test inconclusive), mirroring the observer tests, and only breaks on `Empty`/`Closed`. Pushed in 8d7b26af.

File: crates/governance-store/src/tests.rs [hunk 8]
```diff
@@ -5749,6 +5751,301 @@ mod auto_follow_tests {
...
+                member: new_member_pk,
+                role: GroupMemberRole::Member,
+            },
+        )
+        .unwrap();
+        let content_hash = op.content_hash().unwrap();
+
+        // Observer thread: blocks on `recv()`, then snapshots whether the
+        // op-log already contains our op at the instant the event fires.
+        // `Store` is `Clone` (shared handle to the same backing DB), so the
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5805]: Correct that the barrier only guarantees concurrent start, not that the observer is spinning before `notify`. I am NOT taking the "inject a sync point between the append and the flush" route, though: that means a test-only synchronization hook in the apply hot path, which I want to keep out of production for a test.

Characterization instead (now documented in the test):
- For the FIXED code (persist-then-notify) this is a DETERMINISTIC green — the event is only observable after the append, so the snapshot is always `true` regardless of timing.
- As a RED regression-catcher it is best-effort, but the vacuous-pass is not a practical risk: after the barrier the applier must run the membership mutations AND the op-log append (substantial store work) before reaching `notify`, while the observer only needs to enter a tight spin loop — so it is already polling long before the event fires.
- The structural guarantee is the persist-then-flush ordering visible in `apply_local_signed_group_op` / the namespace apply path itself; the `replayed_group_op_does_not_re_emit` test is fully deterministic.

Expanded the test comments to state this explicitly rather than adding a hot-path hook. Pushed in 8d7b26af.

File: crates/governance-store/src/tests.rs [hunk 9]
```diff
@@ -5749,6 +5751,326 @@ mod auto_follow_tests {
...
+        // that hot path. The structural guarantee is the persist-then-flush
+        // ordering visible in `apply_local_signed_group_op` itself.
+        let barrier = Arc::new(Barrier::new(2));
+        let observer_barrier = Arc::clone(&barrier);
+        let observer = std::thread::spawn(move || {
+            observer_barrier.wait();
+            let deadline = std::time::Instant::now() + std::time::Duration::from_secs(10);
+            loop {
+                if std::time::Instant::now() >= deadline {
+                    return None;
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5835]: Applied — both observers now use `std::thread::yield_now()` instead of `std::hint::spin_loop()`, so a single-core / cooperative scheduler can schedule the applier thread instead of being starved into the 10s deadline. It still polls tightly enough to snapshot near the broadcast send. Pushed in 138470b7.

File: crates/governance-store/src/tests.rs [hunk 10]
```diff
@@ -5749,6 +5751,326 @@ mod auto_follow_tests {
...
+    fn replayed_group_op_does_not_re_emit() {
+        use crate::op_events::{self, OpEvent};
+
+        let mut rng = OsRng;
+        let (store, _gid, gid_bytes, admin_sk, _existing_member_sk) = seed(&mut rng);
+
+        let new_member_sk = PrivateKey::random(&mut rng);
+        let new_member_pk = new_member_sk.public_key();
+
+        let op = SignedGroupOp::sign(
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5900]: Applied your suggestion. The test now subscribes BEFORE the first apply (no subscribe-to-apply gap), runs both applies, then drains once and asserts `MemberAdded` for (gid, member) appears EXACTLY ONCE — the first apply emits it, the replay (already-logged) must not. This eliminates the window and adds a positive control (a mechanism that silently fired zero times now also fails). `Lagged` still panics (a flood that drops counted events is inconclusive, not a false pass). Pushed in 138470b7.

File: crates/governance-store/src/tests.rs [hunk 11]
```diff
@@ -5749,6 +5751,326 @@ mod auto_follow_tests {
...
+            &admin_sk,
+            gid_bytes,
+            vec![],
+            [0u8; 32],
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5904]: Applied your suggestion. The test now subscribes BEFORE the first apply (no subscribe-to-apply gap), runs both applies, then drains once and asserts `MemberAdded` for (gid, member) appears EXACTLY ONCE — the first apply emits it, the replay (already-logged) must not. This eliminates the window and adds a positive control (a mechanism that silently fired zero times now also fails). `Lagged` still panics (a flood that drops counted events is inconclusive, not a false pass). Pushed in 138470b7.

File: crates/governance-store/src/tests.rs [hunk 12]
```diff
@@ -5749,6 +5751,326 @@ mod auto_follow_tests {
...
+                        group_id, member, ..
+                    }) if group_id == gid_bytes && member == new_member_pk => {
+                        return Some(
+                            crate::local_state::op_log_contains_content_hash(
+                                &observer_store,
+                                &gid,
+                                &content_hash,
+                            )
+                            .unwrap(),
+                        );
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5848]: Added a concise top-level `// NOTE: deterministic GREEN / best-effort RED` banner directly above both ordering tests so the characterization is not buried in the body comment.

Declining the two stronger suggestions, with reasoning:
- A test-only sync hook (`#[cfg(test)]` channel/Barrier) injected BETWEEN the append and the flush in `apply_local_signed_group_op`: that puts test scaffolding in a production hot path, which I want to avoid for a best-effort guard.
- `#[ignore]` / `#[cfg_attr(ci, ignore)]`: this would discard the DETERMINISTIC GREEN coverage the test does provide (it reliably proves the fixed persist-then-notify ordering on every run). The RED is best-effort, but disabling the test loses the green signal too.

The deterministic regression guard is structural: the persist-then-flush ordering is visible directly in the apply functions, and `replayed_group_op_does_not_re_emit` is fully deterministic. Pushed in 138470b7.

File: crates/governance-store/src/tests.rs [hunk 13]
```diff
@@ -5749,6 +5751,326 @@ mod auto_follow_tests {
...
+        // regression-catcher it is best-effort: in principle a regression
+        // (notify-then-persist) is only caught if the observer reads the event
+        // inside the notify→persist window. In practice that window is reliably
+        // hit because after the barrier the applier must run the membership
+        // mutations AND the op-log append (substantial store work) before it
+        // reaches `notify`, whereas the observer only needs to enter a tight
+        // spin loop — orders of magnitude less work — so it is already polling
+        // long before the event fires. A FULLY deterministic RED would need a
+        // synchronization point injected BETWEEN the append and the flush in the
+        // production apply path; we deliberately do not add a test-only hook to
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5825]: Added a concise top-level `// NOTE: deterministic GREEN / best-effort RED` banner directly above both ordering tests so the characterization is not buried in the body comment.

Declining the two stronger suggestions, with reasoning:
- A test-only sync hook (`#[cfg(test)]` channel/Barrier) injected BETWEEN the append and the flush in `apply_local_signed_group_op`: that puts test scaffolding in a production hot path, which I want to avoid for a best-effort guard.
- `#[ignore]` / `#[cfg_attr(ci, ignore)]`: this would discard the DETERMINISTIC GREEN coverage the test does provide (it reliably proves the fixed persist-then-notify ordering on every run). The RED is best-effort, but disabling the test loses the green signal too.

The deterministic regression guard is structural: the persist-then-flush ordering is visible directly in the apply functions, and `replayed_group_op_does_not_re_emit` is fully deterministic. Pushed in 138470b7.

File: crates/governance-store/src/tests.rs [hunk 14]
```diff
@@ -5749,6 +5751,340 @@ mod auto_follow_tests {
...
+        let observer_barrier = Arc::clone(&barrier);
+        let observer = std::thread::spawn(move || {
+            observer_barrier.wait();
+            let deadline = std::time::Instant::now() + std::time::Duration::from_secs(10);
+            loop {
+                if std::time::Instant::now() >= deadline {
+                    return None;
+                }
+                match rx.try_recv() {
+                    Ok(OpEvent::MemberAdded {
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5838]: This is a false positive — the comment is anchored to 138470b7, where the subscribe is already BEFORE both applies:

```rust
let mut rx = op_events::subscribe();
apply_local_signed_group_op(&store, &op).unwrap(); // first apply emits
apply_local_signed_group_op(&store, &op).unwrap(); // replay (already logged)
```

(tests.rs in 138470b7, the `let mut rx = op_events::subscribe();` line precedes both `apply_local_signed_group_op` calls.)

So the first emit IS captured, `member_added` counts to 1, and the test passes — `cargo test -p calimero-governance-store` reports `tests::auto_follow_tests::replayed_group_op_does_not_re_emit ... ok` (393 passed; 0 failed). The "`assert_eq!(0, 1)` / broken as written" reading does not match the committed code. No change needed.

File: crates/governance-store/src/tests.rs [hunk 15]
```diff
@@ -5749,6 +5751,340 @@ mod auto_follow_tests {
...
+                    }) if group_id == gid_bytes && member == new_member_pk => {
+                        return Some(
+                            crate::local_state::op_log_contains_content_hash(
+                                &observer_store,
+                                &gid,
+                                &content_hash,
+                            )
+                            .unwrap(),
+                        );
+                    }
```
[the remark this answers is not in the document]
Comment (xilosada) [reply crates/governance-store/src/tests.rs:5849]: Already addressed in 138470b7 (and the prior round, threads ea28329b8e7d / 5138916c7ff2). The exact `// NOTE:` banner you suggest is now directly above the `#[test]`:

```
/// NOTE: this is a DETERMINISTIC GREEN (proves the fixed persist-then-notify
/// ordering) but only a BEST-EFFORT RED (regression detection) — see the
/// `Barrier` comment in the body for why a fully deterministic RED would need
/// a test-only sync hook in the apply hot path, which we deliberately omit.
```

And, as you note, the PR explicitly documents the tradeoff. I am not adding the production-hot-path test hook (a best-effort guard does not justify test scaffolding in the apply path), and the structural guard remains the persist-then-flush ordering in the apply functions plus the fully-deterministic `replayed_group_op_does_not_re_emit`. Considering this resolved.
